### ISSUE TITLE
MYSQL session timeout after 8 hours.

### ISSUE THREAD

**Issue by Amarog** [GriefPrevention/GriefPrevention#1312]
So it seems after 8 hours my connection with the mysql database for griefprevention is lost... 
Is there a way I can make it auto reconnect?
All other plugs work just fine...

when I call for Database inputs ( /claimslist for instance) I see nothing gets updated... 

restarting my server each 8 hours now, but is there another way to handel this?

### What happened:
[14:15:44 INFO]: Amarog issued server command: /claimslist
[14:15:44 INFO]: [GriefPrevention] baed366d-afe5-4e9b-b6a4-c394f6b2ae51 com.mysql.jdbc.exceptions.jdbc4.CommunicationsException: The last packet successfully received from the server was 60,296,942 milliseconds ago.  The last packet sent successfully to the server was 60,296,971 milliseconds ago. is longer than the server configured value of 'wait_timeout'. You should consider either expiring and/or testing connection validity before use in your application, increasing the server configured values for client timeouts, or using the Connector/J connection property 'autoReconnect=true' to avoid this problem.

### What was expected:
Getting aquired blocks

### Steps to reproduce:
__This may include a build schematic, a video, or detailed instructions to help reconstruct the issue__
1. Step 1.
2. Step 2.
3. ???
4. Profit!

### Server and GriefPrevention version:
GriefPrevention version 16.17.1-2b69191

**Comment by Jikoo** [comment]
Frankly, your best bet is to not use MySQL - GP's support for it is janky at best, and won't really be (well) supported until v19. Don't be distracted by the shininess of databases compared to flatfile storage - GP's current implementation kills any potential gain or security.

As an immediate bandaid on your end, you can change your SQL server's `interactive_timeout` and `wait_timeout` properties to prevent the connection being killed. For example, `SET GLOBAL wait_timeout=90000` sets wait timeout to 25 hours. Assuming you schedule daily restarts, the connection should stay open long enough to make it to next restart even if no one causes a GP database query to be made.

The solution on GP's end as far as a simple patch goes would be to run `DatabaseDataStore#refreshDataConnection` prior to every call. GP does already use the auto-reconnect property, but that's historically unreliable and has been deprecated since 2004 or so.

From [MySQL Connector/J](https://dev.mysql.com/doc/connector-j/8.0/en/connector-j-usagenotes-troubleshooting.html#qandaitem-15-1-4):

>Note
Use of the `autoReconnect` option is not recommended because there is no safe method of reconnecting to the MySQL server without risking some corruption of the connection state or database state information. Instead, use a connection pool, which will enable your application to connect to the MySQL server using an available connection from the pool. The `autoReconnect` facility is deprecated, and may be removed in a future release.

**Comment by RoboMWM** [comment]
Yea, I think nouish had similar comments in another issue.

**Comment by Amarog** [comment]
Yeah, but I like the Database, I'm using it to read statistics and throw them online so my players who don't get around all the console commands can see how they are doing...

**Comment by weihao** [comment]
https://github.com/TechFortress/GriefPrevention/blob/6debf368c9823c01fe747d70d5c1845024d86fa1/src/main/java/me/ryanhamshire/GriefPrevention/DatabaseDataStore.java#L542

https://github.com/TechFortress/GriefPrevention/blob/6debf368c9823c01fe747d70d5c1845024d86fa1/src/main/java/me/ryanhamshire/GriefPrevention/DatabaseDataStore.java#L497

Add validation before executing the query.

Or we could catch the exception after the query is failed, and use refreshDataConnection.

https://github.com/TechFortress/GriefPrevention/blob/6debf368c9823c01fe747d70d5c1845024d86fa1/src/main/java/me/ryanhamshire/GriefPrevention/DatabaseDataStore.java#L622

@Jikoo thoughts on this?  we need db support because flat files are just not optimized.

**Comment by Jikoo** [comment]
> @Jikoo thoughts on this? we need db support because flat files are just not optimized.

My thoughts remain the same that they were when I last posted on this issue: https://github.com/TechFortress/GriefPrevention/issues/1312#issuecomment-799470630

I don't think you understand how poorly optimized GP's use of SQL is honestly, there is little to no benefit to using it. Flatfile is unfortunately exactly as (un)reliable and probably just as fast.

**Comment by weihao** [comment]
> > @Jikoo thoughts on this? we need db support because flat files are just not optimized.
> 
> My thoughts remain the same that they were when I last posted on this issue: [#1312 (comment)](https://github.com/TechFortress/GriefPrevention/issues/1312#issuecomment-799470630)
> 
> I don't think you understand how poorly optimized GP's use of SQL is honestly, there is little to no benefit to using it. Flatfile is unfortunately exactly as (un)reliable and probably just as fast.

I have SQL migrations scripts and apps that directly interact with the SQL database. I can't convert them into bash to use flat files, not to mention the flatfiles are not optimized for access.

**Comment by Jikoo** [comment]
> I have SQL migrations scripts and apps that directly interact with the SQL database. I can't convert them into bash to use flat files, not to mention the flatfiles are not optimized for access.

I posted a viable solution for a SQL-side bandaid. You can either use that or submit a PR to fix the issue.
